lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,209 @@
+	// WARNING: The create() function cannot rely on any of the struct attributes being set
+
+	// Set default source if missing
+	if dbPool.Config["source"] == "" {
+		dbPool.Config["source"] = GetStoragePoolMountPoint(dbPool.Name)
+	}
+
+	if !shared.PathExists(dbPool.Config["source"]) {
+		return fmt.Errorf("Source path '%s' doesn't exist", dbPool.Config["source"])
+	}
+
+	// Check that the path is currently empty
+	isEmpty, err := shared.PathIsEmpty(dbPool.Config["source"])
+	if err != nil {
+		return err
+	}
+
+	if !isEmpty {
+		return fmt.Errorf("Source path '%s' isn't empty", dbPool.Config["source"])
+	}
+
+	return nil
+}
+
+func (d *dir) Name() string {
+	return "dir"
+}
+
+func (d *dir) Version() string {
+	return "1"
+}
+
+func (d *dir) Delete(op *operations.Operation) error {
+	// On delete, wipe everything in the directory
+	err := wipeDirectory(d.path)
+	if err != nil {
+		return err
+	}
+
+	// Unmount the path
+	_, err = d.Unmount()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *dir) Mount() (bool, error) {
+	// Check if we're dealing with an external mount
+	if d.config["source"] == d.path {
+		return false, nil
+	}
+
+	// Check if already mounted
+	if sameMount(d.config["source"], d.path) {
+		return false, nil
+	}
+
+	// Setup the bind-mount
+	err := tryMount(d.config["source"], d.path, "none", unix.MS_BIND, "")
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (d *dir) Unmount() (bool, error) {
+	// Check if we're dealing with an external mount
+	if d.config["source"] == d.path {
+		return false, nil
+	}
+
+	// Unmount until nothing is left mounted
+	return forceUnmount(d.path)
+}
+
+func (d *dir) GetResources() (*api.ResourcesStoragePool, error) {
+	// Use the generic VFS resources
+	return vfsResources(d.path)
+}
+
+func (d *dir) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	return ErrNotImplemented
+}
+
+func (d *dir) RenameVolume(volType VolumeType, volName string, newVolName string, op *operations.Operation) error {
+	return ErrNotImplemented
+}
+
+func (d *dir) MigrationType() migration.MigrationFSType {
+	return migration.MigrationFSType_RSYNC
+}
+
+func (d *dir) PreservesInodes() bool {
+	return false
+}
+
+func (d *dir) CreateVolume(volType VolumeType, volID int64, volName string, op *operations.Operation) error {
+	volPath := fmt.Sprintf("%s/%s/%s", d.path, volType, volName)
+	err := os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	err = d.initQuota(volPath, volID)
+	if err != nil {
+		os.RemoveAll(volPath)
+		return err
+	}
+
+	return nil
+}
+
+func (d *dir) CopyVolume(srcPool string, srcVolName string, snapshots []string, volType VolumeType, volID int64, volName string, op *operations.Operation) error {
+	var srcVolPath string
+
+	if shared.IsSnapshot(srcVolName) {
+		srcVolPath = GetStoragePoolVolumeSnapshotMountPoint(srcPool, srcVolName)
+	} else {
+		srcVolPath = GetStoragePoolVolumeMountPoint(srcPool, srcVolName)
+	}
```

Comment:
@stgraber I called this function `CopyVolume` rather than `CopyCustomVolume` as wanted to keep it generic as possible in the driver implementation. However the helper functions used in the previous implementation; `GetStoragePoolVolumeSnapshotMountPoint` and `GetStoragePoolVolumeMountPoint` assume that the volume path includes "custom" and further down `GetStoragePoolVolumeSnapshotMountPoint` assumes "custom-snapshots". 

Is it worth making more generic versions of these helper functions that would accept the `VolumeType` as an argument and build the path based on that, or do you think that copying an "instance" volume would be sufficiently different as to need a separate "CopyInstanceVolume" function?

Reply:
See the comment I left on an earlier commit about `CreateVolumeFromCopy` and the workflow I envision for containers/virtual-machines creation from images through it.

Reply:
Ah and worth noting that VolumeType does differentiate between volumes and snapshots, so we shouldn't need IsSnapshot at all.

Reply:
@stgraber the constants you've added so far for VolumeType do not included the snapshot variants, was it your intention to add more for those variants, such as `VolumeTypeCustomSnapshot`?

Reply:
@stgraber I've solved this with the `GetVolumeMountPoint` function that detects the "/" in the volume name and treats it as a snapshot, so no new constants needed.